lib/reversetunnel/remotesite.go

```diff
@@ -109,39 +126,93 @@ func (s *remoteSite) addConn(conn net.Conn, sshConn ssh.Conn) (*remoteConn, erro
 	return rc, nil
 }
 
+func (s *remoteSite) getLatestTunnelConnection() (services.TunnelConnection, error) {
+	conns, err := s.srv.AccessPoint.GetTunnelConnections(s.domainName)
+	if err != nil {
+		s.Warningf("failed to fetch tunnel statuses: %v", err)
+		return nil, trace.Wrap(err)
+	}
+	var lastConn services.TunnelConnection
+	for i := range conns {
+		conn := conns[i]
+		if lastConn == nil || conn.GetLastHeartbeat().After(lastConn.GetLastHeartbeat()) {
+			lastConn = conn
+		}
+	}
+	if lastConn == nil {
+		return nil, trace.NotFound("no connections found")
+	}
+	return lastConn, nil
+}
+
 func (s *remoteSite) GetStatus() string {
-	s.Lock()
-	defer s.Unlock()
-	diff := time.Now().Sub(s.lastActive)
-	if diff > 2*defaults.ReverseTunnelAgentHeartbeatPeriod {
+	connInfo, err := s.getLatestTunnelConnection()
+	if err != nil {
 		return RemoteSiteStatusOffline
 	}
-	return RemoteSiteStatusOnline
+	if s.isOnline(connInfo) {
+		return RemoteSiteStatusOnline
+	}
+	return RemoteSiteStatusOffline
 }
 
-func (s *remoteSite) setLastActive(t time.Time) {
-	s.Lock()
-	defer s.Unlock()
-	s.lastActive = t
+func (s *remoteSite) copyConnInfo() services.TunnelConnection {
+	s.RLock()
+	defer s.RUnlock()
+	return s.connInfo.Clone()
+}
+
+func (s *remoteSite) registerHeartbeat(t time.Time) {
+	connInfo := s.copyConnInfo()
+	connInfo.SetLastHeartbeat(t)
+	connInfo.SetExpiry(s.clock.Now().Add(defaults.ReverseTunnelOfflineThreshold))
+	err := s.srv.AccessPoint.UpsertTunnelConnection(connInfo)
+	if err != nil {
+		s.Warningf("failed to register heartbeat: %v", err)
+	}
+}
+
+// deleteConnectionRecord deletes connection record to let know peer proxies
+// that this node lost the connection and needs to be discovered
+func (s *remoteSite) deleteConnectionRecord() {
+	s.srv.AccessPoint.DeleteTunnelConnection(s.connInfo.GetClusterName(), s.connInfo.GetName())
 }
 
 func (s *remoteSite) handleHeartbeat(conn *remoteConn, ch ssh.Channel, reqC <-chan *ssh.Request) {
```

Comment:
Comment needed for non-trivial function.

Reply:
:white_check_mark: